### PR TITLE
Basic income enhancements: AGI limit, dollar-range phase-out, refactoring

### LINKED ISSUE #1238: AGI limit for basic income

### LINKED ISSUE #1312: Phase out basic income by a dollar range
We currently offer a phase-out threshold (by filing status) and rate for basic income. Let's also add an option to phase out the basic income by a dollar range, specifying the phase-out end and a toggle to use either the rate or phase-out end.

For example, the paper [Guaranteed Income for the 21st Century](https://drive.google.com/file/d/1UDFPwUYu2Rf4RGgXuOTacmBj2Gt9paAV/view) takes this approach.

### PR DESCRIPTION
**Pull request by MaxGhenis**
Fixes #1238
Fixes #1312

### CODE REVIEW

File: openfisca_us/parameters/contrib/ubi_center/basic_income/agi_limit/in_effect.yaml
Comment (MaxGhenis) [inline openfisca_us/parameters/contrib/ubi_center/basic_income/agi_limit/in_effect.yaml:5]: Is this appropriate?

File: openfisca_us/variables/contrib/ubi_center/basic_income/basic_income.py
Comment (nikhilwoodruff) [inline openfisca_us/variables/contrib/ubi_center/basic_income/basic_income.py:23]: We could use `defined_for = "basic_income_eligible" here and remove this, though I think since there's practically zero performance benefit it's better to use pure OpenFisca formulas.

**Review by nikhilwoodruff — APPROVED** [review]

File: openfisca_us/parameters/contrib/ubi_center/basic_income/agi_limit/in_effect.yaml
Comment (MaxGhenis) [reply openfisca_us/parameters/contrib/ubi_center/basic_income/agi_limit/in_effect.yaml:5]: done

File: openfisca_us/variables/contrib/ubi_center/basic_income/basic_income.py
Comment (MaxGhenis) [reply openfisca_us/variables/contrib/ubi_center/basic_income/basic_income.py:23]: Ah thanks, I switched to `defined_for` for consistency, since it'll improve performance in other cases to adopt the `defined_for = "*_eligible"` pattern.

**Comment by MaxGhenis** [comment]
@nikhilwoodruff I started some enhancements in this PR just before your approval. I'll probably go ahead and merge this but if you have feedback on any new changes please let me know.

File: openfisca_us/variables/contrib/ubi_center/basic_income/basic_income_before_phase_out.py
Comment (nikhilwoodruff) [inline openfisca_us/variables/contrib/ubi_center/basic_income/basic_income_before_phase_out.py:15]: Yeah still not sure about this convention but extremely minor

**Review by nikhilwoodruff — COMMENTED** [review]
Nice, LGTM

**Pull request merged**
